### ISSUE TITLE
Actuator is using CPU while app is doing nothing

### ISSUE THREAD

**Issue by dgsqrs** [spring-projects/spring-boot#11333]
Hi there.

If you add the spring-boot-starter-actuator as dependency, after the first http call, the app will consume 6%-7% CPU forever (on a few threads). Even when there is no active http connection nor application background thread working.

If I remove this dependency, the app is idle at 0% as expected.

**Edit**
Spring Boot version 2.0.0.0M7
Java 8 
[Archive.zip](https://github.com/spring-projects/spring-boot/files/1554553/Archive.zip)

The demo app have been generated with https://start.spring.io/

Steps to reproduce the issue
1) Build the app
2) Launch it using java -jar
3) Monitor CPU usage using htop => Nothing special is seen
4) Open a web browser and browse to localhost:8080 => error page is shown
5) Check CPU usage on htop => 6-7% used endlessly by the app

**Comment by philwebb** [comment]
@dgsqrs What very of Spring Boot are you using? Can you share a small sample that shows the issue?

**Comment by dgsqrs** [comment]
Hi philwebb.
I updated my initial post with the requested information.

**Comment by bclozel** [comment]
I can reproduce this. After the first request, the CPU seems to be busy:

<img width="1090" alt="after-first-request" src="https://user-images.githubusercontent.com/103264/33930863-9badd5ce-dfee-11e7-803f-bba3bd72bba2.png">

VisualVM shows this:
<img width="1095" alt="cpu" src="https://user-images.githubusercontent.com/103264/33930883-ac1dec32-dfee-11e7-83c4-5a14262c7765.png">

This behaviour seems to be related with the new metrics support, since excluding the autoconfiguration with `@SpringBootApplication(exclude = MetricsAutoConfiguration.class)` "fixes" this.

@jkschneider is this related to the GC pause detection or some specific instrumentation? Is the cost constant even when the application is busy or is this just something to keep the VM busy on purpose?

**Comment by jkschneider** [comment]
> Is this related to the GC pause detection
Yes, I believe so.

> Is the cost constant even when the application is busy
Yes.

**Comment by philwebb** [comment]
My initial feeling is that we should set the default to not consume CPU. I expect that GC pause detection is probably quite an advanced problem for most users and by the time that they need it they'll be quite a long way into their journey with metrics. Constant CPU usage, however, is probably going to get noticed.

tldr: I think this should be configurable and by default off. In the docs we should call it out.

**Comment by jkschneider** [comment]
I don't necessarily disagree with the default if the mechanism can't be improved, but this kind of reasoning really troubles me:

> by the time that they need it they'll be quite a long way into their journey with metrics

Unless we compensate for pauses, they are basically looking at garbage. This lures folks into a false sense of security, the "blindfold with a lollipop" state of being.

**Comment by jkschneider** [comment]
I'm pretty confident we can implement a pause detector that trades off pause-length granularity for apparent CPU consumption.

**Comment by osialr** [comment]
I found that setting the system property `LatencyUtils.useActualTime` to `false` disables the `SimplePauseDetector` but keeps the rest of micrometer running.

It switches LatencyUtils into a test mode such that [`TimeServices#sleepNanos`](https://github.com/LatencyUtils/LatencyUtils/blob/master/src/main/java/org/LatencyUtils/TimeServices.java#L59) will call `wait()` on a monitor that never fires.

**Comment by jkschneider** [comment]
> I found that setting the system property `LatencyUtils.useActualTime` to `false`

Wow, the hacking prize goes to you today, @osialr :)

**Comment by snicoll** [comment]
@jkschneider do you have a micrometer issue that we can track?

**Comment by jkschneider** [comment]
Testing the fix this morning. Here is the issue: https://github.com/micrometer-metrics/micrometer/issues/328

**Comment by jkschneider** [comment]
Summarizing the changes coming in Micrometer 1.0.0-rc.6:

The pause detector is configurable to an extent via `MeterRegistry#config()`. The default pause detector will still be a clock-drift detector which needs to wake up periodically to check for drift. By configuring the sleep interval (time between checks), you can control CPU consumption. We've selected a default in rc6 that reduces the CPU consumption to about 0.1% while still providing decent compensation for long pauses.

You can change the intervals on the clock-drift detector like so:

```
@Bean
MeterRegistryConfigurer registryConfig() {
   return r -> r.config().pauseDetector(new ClockDriftPauseDetector(sleepInt, pauseInt));
}
```

We also provide an alternate, noop implementation. Unless you are concerned about even the slightest amount of CPU use, this is not recommended:

```
@Bean
MeterRegistryConfigurer registryConfig() {
   return r -> r.config().pauseDetector(new NoPauseDetector());
}
```

It's possible that we'll find alternate pause detector implementations in the future. Perhaps the JDK will someday provide an event-based interface for GC events. Or we could probably provide a detector limited to GC events by scraping GC logs.

**Comment by bclozel** [comment]
I've tried another round of testing using the latest Spring Boot snapshots; I'm not really seeing an improvement here and even seeing conflicting results.

## Results

Spring Boot and Actuator:
<img width="1295" alt="boot-actuator" src="https://user-images.githubusercontent.com/103264/34945644-41f13aa0-fa04-11e7-8456-b8e4e8fb4dd9.png">

Spring Boot and Actuator, after first request:
<img width="1298" alt="boot-actuator-first-request" src="https://user-images.githubusercontent.com/103264/34945662-4d042ab0-fa04-11e7-8852-de8bc1396d05.png">

Spring Boot and Actuator, `NoPauseDetector`:
<img width="1297" alt="boot-actuator-noop-pause-detector" src="https://user-images.githubusercontent.com/103264/34945669-54f2db4a-fa04-11e7-965f-7ff9da95a924.png">

Spring Boot and Actuator, `NoPauseDetector` after first request:
<img width="1295" alt="boot-actuator-no-pause-detector-first-request" src="https://user-images.githubusercontent.com/103264/34945675-5995f16e-fa04-11e7-9517-8c1cc115521b.png">

Spring Boot without Actuator:
<img width="1212" alt="boot-without-actuator" src="https://user-images.githubusercontent.com/103264/34945680-5e246300-fa04-11e7-8708-567a15724936.png">

Spring Boot with Actuator, but Metrics auto-configuration ignored:
<img width="1295" alt="boot-actuator-metrics-excluded" src="https://user-images.githubusercontent.com/103264/34945690-67b4ef48-fa04-11e7-9872-150334b18f00.png">

## Summary

* "Spring Boot without Actuator" and "Spring Boot with Metrics auto-configuration ignored" are still equivalent.
* We now see the CPU usage even before the first request, which means some initialization code might have changed. CPU usage seems to decrease after the first request.
* `NoPauseDetector` doesn't seem to make a big difference in my tests

There might be something else at play here and we need to look deeper.
Any idea @jkschneider @osialr ?

**Comment by jkschneider** [comment]
Fixed a bug in `micrometer-core` that prevented the configured detector from being used 
 (https://github.com/micrometer-metrics/micrometer/issues/350).

**Comment by philwebb** [comment]
@jkschneider Is there a SNAPSHOT available that we can try? I'd like to test this before you cut RC8

**Comment by jkschneider** [comment]
@philwebb Yes. Snapshots are published on every successful build to [repo.spring.io](http://repo.spring.io/libs-snapshot/io/micrometer/micrometer-core/1.0.0-SNAPSHOT/)

**Comment by philwebb** [comment]
Thanks @jkschneider. Did you have a PR lined up for this as well? I can't remember what we discussed and if the upgrade is easy or not.

**Comment by bclozel** [comment]
I've hacked a draft PR [here](https://github.com/bclozel/spring-boot/tree/gh-11333).

I'm not sure about a few of those changes, especially in our test suite where we were previously counting measures (if I'm not mistaken). But the test suite is green.

**Comment by jkschneider** [comment]
@bclozel Oops, I guess you got a little ahead of me.

I'll be piling more commits into: https://github.com/spring-projects/spring-boot/pull/11776

**Comment by snicoll** [comment]
OK so the update to the snapshot has fixed the issue for me. @dgsqrs if you  can give `2.0.0.BUILD-SNAPSHOT` a try that would be great.
